graph/network/diffusion.go

```diff
@@ -0,0 +1,212 @@
+// Copyright ©2017 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package network
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/mat"
+)
+
+// Diffuse performs a heat diffusion across nodes of the undirected
+// graph described by the given Laplacian using the initial heat distribution,
+// h, according to the Laplacian with a diffusion time of t.
+// The resulting heat distribution is returned, written into the map dst and
+// returned,
+//  d = exp(-Lt)×h
+// where L is the graph Laplacian. Indexing into h and dst is defined by the
+// Laplacian Index field. If dst is nil, a new map is created.
+//
+// Nodes without corresponding entries in h are given an initial heat of zero,
+// and entries in h without a corresponding node in the original graph are
+// not altered when written to dst.
+func Diffuse(dst, h map[int64]float64, by Laplacian, t float64) map[int64]float64 {
+	heat := make([]float64, len(by.Index))
+	for id, i := range by.Index {
+		heat[i] = h[id]
+	}
+	v := mat.NewVecDense(len(heat), heat)
+
+	var m, tl mat.Dense
+	tl.Scale(-t, by)
+	m.Exp(&tl)
+	v.MulVec(&m, v)
+
+	if dst == nil {
+		dst = make(map[int64]float64)
+	}
+	for i, n := range v.RawVector().Data {
+		dst[by.Nodes[i].ID()] = n
+	}
+	return dst
+}
+
+// DiffuseToEquilibrium performs a heat diffusion across nodes of the
+// graph described by the given Laplacian using the initial heat
+// distribution, h, according to the Laplacian until the update function
+//  h_{n+1} = h_n - L×h_n
+// results in a 2-norm update difference within tol, or iter updates have
+// been made.
+// The resulting heat distribution is returned as eq, written into the map dst,
+// and a boolean indicating whether the equilibrium converged to within tol.
+// Indexing into h and dst is defined by the Laplacian Index field. If dst
+// is nil, a new map is created.
+//
+// Nodes without corresponding entries in h are given an initial heat of zero,
+// and entries in h without a corresponding node in the original graph are
+// not altered when written to dst.
+func DiffuseToEquilibrium(dst, h map[int64]float64, by Laplacian, tol float64, iters int) (eq map[int64]float64, ok bool) {
```

Comment:
> There isn't a more efficient way of computing the equilibrium? I thought it was one of the eigenvalues of the Laplacian?

If we consider the transition matrix T and do the iteration `p_{i+1} = T p_i`, then *if* `p_i` converges to `p`, we get `p = Tp`. This means that `p` is an eigenvector corresponding to the eigenvalue 1 and this algorithm is a way of finding it. It should converge quite fast.

The problem is in the "if" above: it may not converge. 1 is always an eigenvalue of T. If all other eigenvalues are smaller than 1 in absolute value and the graph is connected, 1 has multiplicity 1, and this converges for any initial vector. If there are more components, the multiplicity of 1 will be equal to the number of components and the limit may depend on the starting vector. If -1 is an eigenvalue, it may not converge.

Random walk with the transition matrix is useful for finding the eigenvector corrsponding to 1 (which is useful as a centrality measure) without finding the complete spectrum of the matrix, so I think this should stay.

Reply:
Thanks

Reply:
Okay, thanks!